### DISCUSSION TITLE
Is there a way to transform the uploaded image and do not keep the original?

### DISCUSSION THREAD

**Discussion by titogarrido** [directus/directus#8704] [Archive]
Example, upload a big image file, transform resize and optimize it, and remove the original... Is there a way to do that?

**Comment by joselcvarela** [comment]
Hello @titogarrido 
I think this should be possible.
But the transformations will need to be made on inserting the file, because currently the transformations are made on the first read of the asset.

That being said, you can register the `files.upload` and then you will have access to these: 
https://github.com/directus/directus/blob/main/api/src/services/files.ts#L127-L136

You can extract `payload.filename_disk` and do your modifications, in this case you should create a new temp file with the pretended size using sharp (for example), remove the original file and then rename the temp file to the original one (`filename_disk`).

**Reply by titogarrido** [reply]
Thank you! I will try that!

**Reply by christian-fuss** [reply]
I am currently trying to do the same, as a complete directus newbie. Have you managed to implement this? Have you done everything on the filesystem level or did you go through the FilesService? It's so sad, that the documentation for developers is so terribly bad, I am currently reading through the directus sources trying to find out, how this could be accomplished in a clean manner. I would be very grateful for any help. Thanks in advance, Christian

**Reply by NilsBaumgartner1994** [reply]
Having the same wish/problem here. Having a 20mb image file is not nice. I will post a solutionin the upcoming days. Seems pretty straight forward for me using the hooks.

**Reply by rijkvanzanten** [reply]
> Having a 20mb image file is not nice.

I disagree! I would personally prefer having the original file stored in the platform, and relying on transformations when using the file to render it at appropriate sizes/compression/resolution. Storing a compressed / smaller original means you can never scale up when the requirement arises, which can cause headaches in the future. I remember when retina / 4k screens just came out, and suddenly everything needed a `@2x` and later an `@3x`. Only the websites I built where the originals were saved along with the content were able to retroactively update, everything else was stuck with the "optimized" assets.

**Reply by christian-hackyourshack** [reply]
I see Rijk's point, but I also know the other side, especially if you work with user-generated content. So I wrote a hook, and it reduces our storage problems a lot. If you want to try it or tweak, go ahead (I am also happy about feedback, since I am still very new to directus internals):

```js
import { defineHook } from "@directus/extensions-sdk";
import {
    PresetFormats,
    Transformation,
    TransformationParams,
    TransformationPreset,
} from "./transform";

export default defineHook(({ action }, { services, env }) => {
    const { AssetsService, FilesService } = services;
    const quality = env.EXTENSIONS_REDUCE_ON_UPLOAD_QUALITY ?? 90;
    const maxSize = env.EXTENSIONS_REDUCE_ON_UPLOAD_MAXSIZE ?? 1920;

    action("files.upload", async ({ payload, key }, context) => {
        if (!payload.optimized) {
            const transformation = getTransformation(payload.type, quality, maxSize);
            if (transformation !== undefined) {
                const serviceOptions = { ...context, knex: context.database };
                const assets = new AssetsService(serviceOptions);
                const files = new FilesService(serviceOptions);

                const { stream, stat } = await assets.getAsset(key, transformation);
                if (stat.size < payload.filesize) {
                    await sleep(500); // Why is this required?
                    await files.uploadOne(stream, { ...payload, optimized: true }, key);
                }
            }
        }
    });
});

function getTransformation(
    type: string,
    quality: number,
    maxSize: number
): TransformationParams | TransformationPreset | undefined {
    const format = type.split("/")[1] ?? "";
    if (PresetFormats.includes(format)) {
        const transforms: Transformation[] = [["withMetadata"]];
        if (format === "jpeg") {
            transforms.push(["jpeg", { progressive: true }]);
        }
        return {
            format,
            quality,
            width: maxSize,
            height: maxSize,
            fit: "inside",
            withoutEnlargement: true,
            transforms,
        };
    }
    return undefined;
}

async function sleep(ms: number) {
    return new Promise(r => setTimeout(r, ms));
}
```

**Reply by martabitbrain** [reply]
I've had this issue too with user-generated content. 

Just uploaded a version of @christian-hackyourshack 's code in the form of a working directus hook extension, just in case someone stumbles on this discussion again: 

https://github.com/martabitbrain/directus-extension-sane-image-size

Maybe it needs some further tweaks, this is my first attempt at making an extension.

Cheers! And thank you Christian for the code <3.

**Reply by christian-hackyourshack** [reply]
Thank you @martabitbrain for upgrading the code to work with the latest version of Directus and making it configurable. Great to see, that it is used somewhere else.

**Reply by kolosseo** [reply]
> Thank you @martabitbrain for upgrading the code to work with the latest version of Directus and making it configurable. Great to see, that it is used somewhere else.

I think the old extension died, anyway, I think can be useful to say that I also built an extension [Image Upload Resizer](https://github.com/kolosseo/directus-extension-image-upload-resizer) taking inspiration by your code.  
Actually it reached version 2.2 and it manages JPEG, PNG, WebP, AVIF, TIFF, also HEIC and HEIF from Apple devices.

It is installable straight from the Directus Marketplace (you need to set `MARKETPLACE_TRUST=all` previously).

NPM: https://www.npmjs.com/package/directus-extension-image-upload-resizer

GitHub: https://github.com/kolosseo/directus-extension-image-upload-resizer

Hope this can be helpful to someone.

**Comment by NilsBaumgartner1994** [comment]
@martabitbrain woudl be nice to see an upvote on the newly create Feature Request: https://github.com/directus/directus/discussions/22622
@christian-hackyourshack

**Reply by martabitbrain** [reply]
Yes, of course! Done!

**Comment by rowild** [comment]
What happens to those who need large images? Photographers, architects, anybody who wants to provide high resolution print-ready stuff?

**Reply by martabitbrain** [reply]
That would be not a problem since the feature should be opt-in (It should be able to be activated on selected collections, or even specific fiedls). The extension I fixed does not support it at the moment because it was a emergency measure I needed😊. The images were so big that transformations performed on directus were failing.

**Reply by rijkvanzanten** [reply]
Storage is generally cheap, so I'm often erring on the side of the leaving the originals in place, and operate against thumbnails. That being said, implementing some "default" transformation on upload to make sure images don't exceed a maximum size configured by the admin could be an interesting enhancement of the system 🙂

**Comment by NilsBaumgartner1994** [comment]
I totally agree with this. I would also be very happy that my admins or customers don’t upload their 8k image of a salad.

**Reply by rijkvanzanten** [reply]
Are you imagining you'd _prevent_ them from uploading it in the first place, or would you expect it to finish the upload, then force resize it to be smaller, and leave it in?

**Reply by Dominic-Marcelino** [reply]
Personally I have a use-case for both. Not directly the scope of this discussion but I’d love to have a filter to fully prevent an upload in certain cases.

As a Workaround it probably could be solved by a users-value based permission check but a filter would be a great addition too ☺️

**Reply by kolosseo** [reply]
> Personally I have a use-case for both. Not directly the scope of this discussion but I’d love to have a filter to fully prevent an upload in certain cases.

To filter large files you could use the env  variable [FILES_MAX_UPLOAD_SIZE](https://directus.com/docs/configuration/files#upload-limits) i.e. `FILES_MAX_UPLOAD_SIZE=8mb`.

**Comment by NilsBaumgartner1994** [comment]
Both approaches may have their own optimal use case. I think it would be
neat to have an upload transformer to automatically reduce the size.


Rijk van Zanten ***@***.***> schrieb am Di. 18. Juni 2024 um
20:18:
